atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -80,117 +65,142 @@ private CassandraTimestampBoundStore(CassandraClientPool clientPool) {
             result = client.cas(
-                    getRowName(),
+                    ROW_NAME_BYTE_BUFFER,
                     AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName(),
-                    oldVal == null ? ImmutableList.of() : ImmutableList.of(makeColumn(oldVal)),
-                    ImmutableList.of(makeColumn(newVal)),
+                    makeListOfColumnsFromEntry(entryInDb),
+                    makeListOfColumnsFromEntry(newEntry),
                     ConsistencyLevel.SERIAL,
                     ConsistencyLevel.EACH_QUORUM);
+            return result;
         } catch (Exception e) {
-            log.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
-            DebugLogger.logger.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
+            CassandraTimestampUtils.logUpdateUncheckedException(entryInDb, newEntry);
             throw Throwables.throwUncheckedException(e);
         }
-        if (!result.isSuccess()) {
-            String msg = "Unable to CAS from {} to {}."
-                    + " Timestamp limit changed underneath us (limit in memory: {}, stored in DB: {})."
-                    + " This may indicate that another timestamp service is running against this cassandra keyspace."
-                    + " This is likely caused by multiple copies of a service running without a configured set of"
-                    + " leaders or a CLI being run with an embedded timestamp service against an already running"
-                    + " service.";
-            MultipleRunningTimestampServiceError err = new MultipleRunningTimestampServiceError(
-                    String.format(replaceBracesWithStringFormatSpecifier(msg),
-                            oldVal,
-                            newVal,
-                            currentLimit,
-                            getCurrentTimestampValues(result)));
-            log.error(msg, oldVal, newVal, currentLimit, getCurrentTimestampValues(result), err);
-            DebugLogger.logger.error(msg, oldVal, newVal, currentLimit, getCurrentTimestampValues(result), err);
-            DebugLogger.logger.error("Thread dump: {}", ThreadDumps.programmaticThreadDump());
-            throw err;
+    }
+
+    private void setCurrentLimit(String type, TimestampBoundStoreEntry newEntry) {
+        checkLimitNotDecreasing(newEntry);
+        currentLimit = newEntry.getTimestampOrInitialValue();
+        DebugLogger.logger.info("{} Setting cached timestamp limit to {}.", type, currentLimit);
+    }
+
+    private void retryCasIfMatchingId(Client client, TimestampBoundStoreEntry newEntry, CASResult result) {
+        TimestampBoundStoreEntry entryInDb = TimestampBoundStoreEntry.createFromCasResult(result);
+        if (entryInDb.idMatches(id)) {
+            setCurrentLimit("[CAS]", entryInDb);
+            entryInDb = TimestampBoundStoreEntry.create(currentLimit, id);
+            cas(client, entryInDb, newEntry);
```

Comment:
In theory we could end up with a long loop. But only if someone actually keeps writing new increasing bounds with our ID that are also smaller than the bound that we actually want to write. Since that happening once is a one in several months event, it happening repeatedly enough times for us to notice is not a real concern. 

Feel free to push back if you don't agree.